### DISCUSSION TITLE
Where are "Restore Ban"s getting the list of bans to restore?

### DISCUSSION THREAD

**Discussion by brianjmurrell** [fail2ban/fail2ban#3821] [Q&A] [answered]
I have log entries on a restart of the form `NOTICE  [jail] Restore Ban <ip>`.

Where is the information (i.e. the IP address) that that ban is restoring coming from?

**Comment by sebres** [comment, marked as answer]
They are stored in the sqlite database, see https://github.com/fail2ban/fail2ban/blob/54c0effceb998b73545073ac59c479d9d9bf19a4/config/fail2ban.conf#L64-L70

**Reply by brianjmurrell** [reply]
This is as I suspected.  Can you be more detailed though, such as which table(s) in the database does this info come from?

I ask because I am fairly sure I have inspected some of these restored bans and cannot find any presence of them in the database.  But I am sure fail2ban is not just making them up so I must just not be looking in the right place.

**Reply by brianjmurrell** [reply]
I'm seeing there are a couple of tables that might be relevant here:

```
CREATE TABLE bans(jail TEXT NOT NULL, ip TEXT, timeofban INTEGER NOT NULL, bantime INTEGER NOT NULL, bancount INTEGER NOT NULL default 1, data JSON, FOREIGN KEY(jail) REFERENCES jails(name) );
CREATE TABLE bips(ip TEXT NOT NULL, jail TEXT NOT NULL, timeofban INTEGER NOT NULL, bantime INTEGER NOT NULL, bancount INTEGER NOT NULL default 1, data JSON, PRIMARY KEY(ip, jail), FOREIGN KEY(jail) REFERENCES jails(name) );
```

I can find references to a restored ban in's `ip` (address) the latter but nothing in the former.

Ultimately though, bans are static and persistent on a device (i.e. my Internet gateway/router) different from where fail2ban runs and stopping fail2ban does not remove the blocks on that device.  In such a case, I probably don't want to be restoring them, do I?

**Reply by sebres** [reply]
It is the table bips. In particular function getCurrentBans does it and here is the statement:
https://github.com/fail2ban/fail2ban/blob/54c0effceb998b73545073ac59c479d9d9bf19a4/fail2ban/server/database.py#L770

> Ultimately though, bans are static and persistent on a device (i.e. my Internet gateway/router) different from where fail2ban runs and stopping fail2ban does not remove the blocks on that device. In such a case, I probably don't want to be restoring them, do I?

In normal case a stop of jail or fail2ban removes everything (e. g. does flush or mass unban), so a restore process is expected after start.
If your action doesn't need that due to persistence, just set `norestored = 1` to the action definition. See #1669 for details.
Just in this case it'd never do an unban for that tickets (they are quasi not banned anymore from point of view of the jail).
Alternatively you can do something like `if [ '<restored>' = '1' ]; then exit 0; fi;` in the `actionban`, so it'll be restored and so remains banned, but would not execute further commands in `actionban`.

**Reply by brianjmurrell** [reply]
That's all great info.

In retrospect, given that I am just trying to remove bans that were added mistakenly, I think ultimately I just want to remove them from `bips` so that restoring in the future won't re-add them.  For audit purposes though, I think I'd rather they stayed in the database but were just "deactivated" for the purposes of future ban restoration.  I suppose setting `bantime` to 0 would achieve that?

**Reply by sebres** [reply]
> I think ultimately I just want to remove them from bips so that restoring in the future won't re-add them

Take a look at `fail2ban-client --help | grep unban`, particularly:
```bash
fail2ban-client unban $ip1 $ip2 ... $ipN; # unban IPs in all jails
fail2ban-client set $jail unbanip $ip1 $ip2 ... $ipN; # unban IPs in single jail
```
(both commands will also remove them from database)

> I suppose setting bantime to 0 would achieve that?

No, it'd just unban IP immediately. If it is needed, it is OK, but this may also bother sometimes... for instance if an IP get banned and hereafter some (retarded) logs causes a second ban, in banned state one'd simply get already banned message in fail2ban log, however by immediate unban (bantime = 0), it'd cause that `actionban` will be invoked second time.
